vulnerabilities/importers/project_kb_msr2019.py

```diff
@@ -0,0 +1,96 @@
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# When you publish or redistribute any data created with VulnerableCode or any VulnerableCode
+# derivative work, you must accompany this data with the following acknowledgment:
+#
+#  Generated with VulnerableCode and provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  VulnerableCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  VulnerableCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/vulnerablecode/ for support and download.
+
+import csv
+import dataclasses
+import urllib.request
+
+# Reading CSV file from  a url using `requests` is bit too complicated.
+# Use `urllib.request` for that purpose. Need `requests` because making
+# a HEADER request using `urllib.request` is too complicated.
+import requests
+from packageurl import PackageURL
+
+
+from vulnerabilities.data_source import Advisory
+from vulnerabilities.data_source import DataSource
+from vulnerabilities.data_source import Reference
+from vulnerabilities.data_source import DataSourceConfiguration
+
+
+@dataclasses.dataclass
+class ProjectKBDataSourceConfiguration(DataSourceConfiguration):
+    etag: dict
+
+
+class ProjectKBMSRDataSource(DataSource):
+
+    CONFIG_CLASS = ProjectKBDataSourceConfiguration
+
+    url = "https://raw.githubusercontent.com/SAP/project-kb/master/MSR2019/dataset/vulas_db_msr2019_release.csv"  # nopep8
+
+    def updated_advisories(self):
+        # etag are like hashes of web responses. We maintain
+        # (url, etag) mappings in the DB. `create_etag`  creates
+        # (url, etag) pair. If a (url, etag) already exists then the code
+        # skips processing the response further to avoid duplicate work
+        if self.create_etag(self.url):
+            raw_data = self.fetch()
+            advisories = self.to_advisories(raw_data)
+            return self.batch_advisories(advisories)
+
+        return []
+
+    def create_etag(self, url):
+        etag = requests.head(url).headers.get("ETag")
+        if not etag:
+            return True
+
+        elif url in self.config.etag:
+            if self.config.etag[url] == etag:
+                return False
+
+        self.config.etag[url] = etag
+        return True
+
+    def fetch(self):
+        response = urllib.request.urlopen(self.url)
+        lines = [l.decode("utf-8") for l in response.readlines()]
+        return csv.reader(lines)
+
+    @staticmethod
+    def to_advisories(csv_reader):
+        # Project KB MSR csv file has no header row
+        all_advsiories = []
+        for row in csv_reader:
+            vuln_id, proj_home, fix_commit, _ = row
```

Comment:
What's `_` here?

Reply:
https://github.com/SAP/project-kb/blob/master/MSR2019/dataset/vulas_db_msr2019_release.csv

It's `pos` for all rows. No idea what it means.

Reply:
Note that FWIW, you can always write:
`for vuln_id, proj_home, fix_commit, _ in csv_reader:`